### PR TITLE
NLL: Improve move error loop detection (was "First shot at #54015")

### LINKED ISSUE #54015: NLL: Improve move error loop detection
In #53995, errors were deduplicated and generally this kept errors with the improved "previous iteration of loop" diagnostic:

```
LL |         while true { while true { while true { x = y; x.clone(); } } }
   |                                                    ^ value moved here in previous iteration of loop
```
However, in one case, [liveness-move-in-while.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/liveness/liveness-move-in-while.rs), the remaining error did not have the improved loop diagnostic.

From @nikomatsakis, in a [comment from that PR (context mine)](https://github.com/rust-lang/rust/pull/53995/files#r215739204):

> actually, I think this [the diagnostic with loop message being missing] is an improvement — or rather it would be if the previous message mentioned the loop. Basically these seem like duplicate errors. I'm not sure why the "borrow of moved value" doesn't mention the loop, though: it seems like a failure in our loop printing heuristic, which is not very good. I think we ought to be walking from the point of the move forward (but ignoring backedges) — if we encounter the use point, then there is no loop, but otherwise we can say "in previous iteration of loop"... or something like that.

### PR DESCRIPTION
**Pull request by blitzerr**
Closes #54015

### CODE REVIEW

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 1]
```diff
@@ -105,7 +105,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             );
 
             let mut is_loop_move = false;
-            for moi in &mois {
+            for (moi, _is_back_edge) in &mois_is_back_edge_pair {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: Nit: no need for `_` in the name here =)

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 2]
```diff
@@ -49,10 +49,10 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             .or_else(|| self.borrow_spans(span, context.loc));
         let span = use_spans.args_or_use();
 
-        let mois = self.get_moved_indexes(context, mpi);
-        debug!("report_use_of_moved_or_uninitialized: mois={:?}", mois);
+        let mois_is_back_edge_pair = self.get_moved_indexes(context, mpi);
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: Nit: I think I'd call this just something like `move_sites` or something

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 3]
```diff
@@ -554,19 +554,25 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         err
     }
 
-    fn get_moved_indexes(&mut self, context: Context, mpi: MovePathIndex) -> Vec<MoveOutIndex> {
+    fn get_moved_indexes(&mut self, context: Context, mpi: MovePathIndex) ->
+        Vec<(MoveOutIndex, bool)> {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: Nit: maybe we should introduce a

```rust
struct MoveSite {
    /// Index of the "move out" that we found. The `MoveData` can
    /// then tell us where the move occurred and so forth.
    moi: MoveOutIndex,

    /// True if we traversed a back edge while walking from the point
    /// of error to the move site.
    traversed_back_edge: bool,
}
```

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 4]
```diff
@@ -554,19 +554,25 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
         let mir = self.mir;
 
         let mut stack = Vec::new();
-        stack.extend(mir.predecessor_locations(context.loc));
+        stack.extend(mir.predecessor_locations(context.loc)
+                     .map(|predecessor| {
+                     // for each of the context's predecessors, if the context also dominates the
+                     // predecessor, then it means there is a back-edge and we try capture that in
+                     // the second field of the tuple.
+                     (predecessor, context.loc.dominates(predecessor, &self.dominators))}));
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: Nit: I'd make a variable here to make this more self-documenting.

```rust
let is_back_edge = context.loc.dominates(predecessor, &self.dominators);
(predecessor,  is_back_edge)
```

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 5]
```diff
@@ -622,7 +628,14 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 continue 'dfs;
             }
 
-            stack.extend(mir.predecessor_locations(l));
+            stack.extend(mir.predecessor_locations(l)
+                         .map(|predecessor| {
+                         // for each of l's predecessors, if 'l' is also a dominant, then that
+                         // implies that there is a back-edge from l to the predecessor and hence
+                         // we add the tuple (predecessor, true) or else (predecessor, false).
+                         (predecessor, l.dominates(predecessor, &self.dominators))
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: The problem is that we want this boolean to be true if we traversed a back edge **at any point along the path**.  So something like this:

```rust
let is_back_edge = context.loc.dominates(predecessor, &self.dominators);
(predecessor,  l.1 || is_back_edge)
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 6]
```diff
@@ -30,6 +30,17 @@ use dataflow::move_paths::indexes::MoveOutIndex;
...
 
+#[derive(Debug, Clone)]
+struct MoveSite {
+    /// Index of the "move out" that we found. The `MoveData` can
+    /// then tell us where the move occurred.
+    moi: MoveOutIndex,
+
+    /// True if we traversed a back edge while walking from the point
+    /// of error to the move site.
+    traversed_back_edge: bool,
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: Nice.

File: src/test/ui/liveness/liveness-move-in-while.stderr [hunk 7]
```diff
@@ -1,21 +1,11 @@
-error[E0382]: use of moved value: `y`
-  --> $DIR/liveness-move-in-while.rs:17:24
+error[E0382]: borrow of moved value: `y`
+  --> $DIR/liveness-move-in-while.rs:18:24
    |
 LL |         println!("{}", y); //~ ERROR use of moved value: `y`
-   |                        ^ value used here after move
-LL |         while true { while true { while true { x = y; x.clone(); } } }
-   |                                                    - value moved here
+   |                        ^ value moved here in previous iteration of loop
```
Comment (nikomatsakis) [inline src/test/ui/liveness/liveness-move-in-while.stderr]: Huh, interesting. This doesn't quite look right, since this location is not a move.

Comment (nikomatsakis) [reply src/test/ui/liveness/liveness-move-in-while.stderr]: Is this the current output?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Updates look good. It'd be nice, if you're up for it, to separate the "rustfmt run" into its own commit (I usually do this by rebasing and "inserting" a rustfmt commit before the "meat"). 

The main concern though is that the output looks wrong; but then I see there's a travis failure, maybe it just needs to be updated?

**Comment by nikomatsakis** [comment]
Ah, no, the travis failure is that the `//~ ERROR` annotations in the test are out of date:

```
[00:52:23] ---- [ui] ui/liveness/liveness-move-in-while.rs stdout ----
[00:52:23] 
[00:52:23] error: /checkout/src/test/ui/liveness/liveness-move-in-while.rs:18: unexpected error: '18:24: 18:25: borrow of moved value: `y` [E0382]'
[00:52:23] 
[00:52:23] error: /checkout/src/test/ui/liveness/liveness-move-in-while.rs:18: expected error not found: use of moved value: `y`
[00:52:23] 
[00:52:23] error: /checkout/src/test/ui/liveness/liveness-move-in-while.rs:19: expected error not found: use of moved value: `y`
```

File: src/test/ui/liveness/liveness-move-in-while.stderr (hunk 7 above)
Comment (nikomatsakis) [reply src/test/ui/liveness/liveness-move-in-while.stderr]: I expect to see "value used here after move", and the "value moved here..." annotation somewhere else.

**Comment by blitzerr** [comment]
joydeep$ rustc +stage2 src/test/ui/liveness/liveness-move-in-while.rs
error[E0382]: borrow of moved value: `y`
  --> src/test/ui/liveness/liveness-move-in-while.rs:18:24
   |
18 |         println!("{}", y); //~ ERROR borrow of moved value: `y`
   |                        ^ value moved here in previous iteration of loop
   |
   = note: move occurs because `y` has type `std::boxed::Box<isize>`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.

**Comment by blitzerr** [comment]
That is the current error when we compile the file.

**Comment by blitzerr** [comment]
```rust
         let move_site_vec = self.get_moved_indexes(context, mpi);
         debug!(
-            "report_use_of_moved_or_uninitialized: mois={:?}",
+            "report_use_of_moved_or_uninitialized: vec<mois,back_edge>={:?}",
             move_site_vec
         );

-        let mois = move_site_vec.clone().into_iter().map(|x| x.moi).collect();
+        let mois = move_site_vec
+            .clone()
+            .into_iter()
+            .map(|x| {
+                let location = self.move_data.moves[x.moi].source;
+                let span = self.mir.source_info(location).span;
+                debug!("moi {:?}: {:?}", x.moi, span);
+                x.moi
+            })
+            .collect();
         if move_site_vec.is_empty() {
             let root_place = self.prefixes(&place, PrefixSet::All).last().unwrap();

@@ -137,12 +146,19 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     ""
                 };

-                if move_site.traversed_back_edge {
+                // if move_site.traversed_back_edge {
+                //     is_loop_move = true;
+                //     err.span_label(
+                //         span,
+                //         format!("value moved{} here in previous iteration of loop", move_msg),
+                //     );
+                // }
+                if span == move_span {
+                    is_loop_move = true;
                     err.span_label(
                         span,
                         format!("value moved{} here in previous iteration of loop", move_msg),
                     );
-                    is_loop_move = true;
                 } else {
                     err.span_label(move_span, format!("value moved{} here", move_msg));
                     move_spans.var_span_label(&mut err, "variable moved due to use in closure");
```

**Comment by blitzerr** [comment]
@nikomatsakis 

This change is basically the same error reporting with the back-edge calculation done (but not used).
Compiled with stage 2 this is the output
```shell
>$ rustc +stage2 ./src/test/ui/liveness/liveness-move-in-while.rs
error[E0382]: borrow of moved value: `y`
  --> ./src/test/ui/liveness/liveness-move-in-while.rs:18:24
   |
18 |         println!("{}", y); //~ ERROR borrow of moved value: `y`
   |                        ^ value moved here in previous iteration of loop
   |
   = note: move occurs because `y` has type `std::boxed::Box<isize>`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
```
Compiled with stage 1, this is the output:
```shell
>$ rustc +stage1 ./src/test/ui/liveness/liveness-move-in-while.rs
error[E0382]: borrow of moved value: `y`
  --> ./src/test/ui/liveness/liveness-move-in-while.rs:18:24
   |
18 |         println!("{}", y); //~ ERROR borrow of moved value: `y`
   |                        ^ value borrowed here after move
19 |         while true { while true { while true { x = y; x.clone(); } } }
   |                                                    - value moved here
   |
   = note: move occurs because `y` has type `std::boxed::Box<isize>`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
```

**Review by davidtwco — COMMENTED** [review]
This looks really great! Just a couple small things I noticed when taking a look at this.

**Comment by davidtwco** [comment]
I couldn't this to the above review below since you didn't change these lines, but could you change the error message in the below snippet to have a comma so all of the "iteration of loop" messages are consistent? (I'm pretty sure this still makes sense gramatically)

ie. `borrow captured here by closure in later iteration of loop` becomes `borrow captured here by closure, in later iteration of loop` 

https://github.com/rust-lang/rust/blob/c222479c6feb036e5e5198e068aa9824d855ffa4/src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs#L48-L52

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 8]
```diff
@@ -733,29 +759,32 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
 
         let mut visited = FxHashSet();
         let mut result = vec![];
 
-        'dfs: while let Some(l) = stack.pop() {
+        'dfs: while let Some((location, is_back_edge)) = stack.pop() {
             debug!(
                 "report_use_of_moved_or_uninitialized: current_location={:?}",
-                l
+                location
```
[the remark this answers is not in the document]
Comment (blitzerr) [reply src/librustc_mir/borrow_check/error_reporting.rs]: Thank you. That is a good point.

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 9]
```diff
@@ -733,29 +759,32 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         err
     }
 
-    fn get_moved_indexes(&mut self, context: Context, mpi: MovePathIndex) -> Vec<MoveOutIndex> {
+    fn get_moved_indexes(&mut self, context: Context, mpi: MovePathIndex) -> Vec<MoveSite> {
         let mir = self.mir;
 
         let mut stack = Vec::new();
-        stack.extend(mir.predecessor_locations(context.loc));
+        stack.extend(mir.predecessor_locations(context.loc).map(|predcessor| {
```
[the remark this answers is not in the document]
Comment (blitzerr) [reply src/librustc_mir/borrow_check/error_reporting.rs]: My lack of attention, no other reason :)

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 10]
```diff
@@ -134,8 +152,16 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                         format!("value moved{} here in previous iteration of loop", move_msg),
```
[the remark this answers is not in the document]
Comment (blitzerr) [reply src/librustc_mir/borrow_check/error_reporting.rs]: sure

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**